ICE: `not implemented: SizeOverflow([u8; usize::MAX])`

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.
http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/
-->

### Code

```Rust
mod assert {
    use std::mem::{Assume, BikeshedIntrinsicFrom};

    pub fn is_maybe_transmutable<Src, Dst>()
    where
        Dst: BikeshedIntrinsicFrom<Src, Context>,
    {
    }
}

fn should_pad_explicitly_packed_field() {
    pub union Uninit {
        _a: [u8; usize::MAX],
    }

    #[repr(C)]
    struct ExplicitlyPadded(Uninit);

    assert::is_maybe_transmutable::<ExplicitlyPadded, ()>();
}

```


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.74.0-nightly (59a829484 2023-08-30)
binary: rustc
commit-hash: 59a8294849358a878a72358aa6d5fe5b9d312867
commit-date: 2023-08-30
host: x86_64-unknown-linux-gnu
release: 1.74.0-nightly
LLVM version: 17.0.0
```

### Error output

```
error[E0412]: cannot find type `Context` in this scope
 --> 943BA15EF022EEDE9240ADD0D6DECED918AC456DCC6493A4A1A05A66F13CEA65.rs:6:41
  |
6 |         Dst: BikeshedIntrinsicFrom<Src, Context>,
  |                                         ^^^^^^^ not found in this scope
  |
help: consider importing this struct
  |
2 +     use std::task::Context;
  |

warning: unused import: `Assume`
 --> 943BA15EF022EEDE9240ADD0D6DECED918AC456DCC6493A4A1A05A66F13CEA65.rs:2:20
  |
2 |     use std::mem::{Assume, BikeshedIntrinsicFrom};
  |                    ^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

error[E0601]: `main` function not found in crate `943BA15EF022EEDE9240ADD0D6DECED918AC456DCC6493A4A1A05A66F13CEA65`
  --> 943BA15EF022EEDE9240ADD0D6DECED918AC456DCC6493A4A1A05A66F13CEA65.rs:20:2
   |
20 | }
   |  ^ consider adding a `main` function to `943BA15EF022EEDE9240ADD0D6DECED918AC456DCC6493A4A1A05A66F13CEA65.rs`

error[E0658]: use of unstable library feature 'transmutability'
 --> 943BA15EF022EEDE9240ADD0D6DECED918AC456DCC6493A4A1A05A66F13CEA65.rs:2:20
  |
2 |     use std::mem::{Assume, BikeshedIntrinsicFrom};
  |                    ^^^^^^
  |
  = note: see issue #99571 <https://github.com/rust-lang/rust/issues/99571> for more information

error[E0658]: use of unstable library feature 'transmutability'
 --> 943BA15EF022EEDE9240ADD0D6DECED918AC456DCC6493A4A1A05A66F13CEA65.rs:2:28
  |
2 |     use std::mem::{Assume, BikeshedIntrinsicFrom};
  |                            ^^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #99571 <https://github.com/rust-lang/rust/issues/99571> for more information

error[E0658]: use of unstable library feature 'transmutability'
 --> 943BA15EF022EEDE9240ADD0D6DECED918AC456DCC6493A4A1A05A66F13CEA65.rs:6:14
  |
6 |         Dst: BikeshedIntrinsicFrom<Src, Context>,
  |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #99571 <https://github.com/rust-lang/rust/issues/99571> for more information
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```

thread 'rustc' panicked at 'not implemented: SizeOverflow([u8; usize::MAX])', compiler/rustc_transmute/src/layout/tree.rs:199:24
stack backtrace:
   0:     0x7f04de162b61 - std::backtrace_rs::backtrace::libunwind::trace::he648b5c8dd376705
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x7f04de162b61 - std::backtrace_rs::backtrace::trace_unsynchronized::h5da3e203eef39e9f
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f04de162b61 - std::sys_common::backtrace::_print_fmt::h8d28d3f20588ae4c
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/sys_common/backtrace.rs:65:5
   3:     0x7f04de162b61 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hd9a5b0c9c6b058c0
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x7f04de1c977f - core::fmt::rt::Argument::fmt::h0afc04119f252b53
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/fmt/rt.rs:138:9
   5:     0x7f04de1c977f - core::fmt::write::h50b1b3e73851a6fe
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/fmt/mod.rs:1094:21
   6:     0x7f04de1554a7 - std::io::Write::write_fmt::h184eaf275e4484f0
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/io/mod.rs:1714:15
   7:     0x7f04de162975 - std::sys_common::backtrace::_print::hf58c3a5a25090e71
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x7f04de162975 - std::sys_common::backtrace::print::hb9cf0a7c7f077819
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x7f04de165753 - std::panicking::default_hook::{{closure}}::h066adb2e3f3e2c07
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/panicking.rs:269:22
  10:     0x7f04de1654e4 - std::panicking::default_hook::h277fa2776900ff14
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/panicking.rs:288:9
  11:     0x7f04e146756b - <rustc_driver_impl[10725d833993dc31]::install_ice_hook::{closure#0} as core[f12ae36cc2e1ecf0]::ops::function::FnOnce<(&core[f12ae36cc2e1ecf0]::panic::panic_info::PanicInfo,)>>::call_once::{shim:vtable#0}
  12:     0x7f04de165f7e - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h09cad52ea08435f2
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/alloc/src/boxed.rs:2007:9
  13:     0x7f04de165f7e - std::panicking::rust_panic_with_hook::hceaf38da6d9db792
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/panicking.rs:709:13
  14:     0x7f04de165d07 - std::panicking::begin_panic_handler::{{closure}}::h2bce3ed2516af7df
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/panicking.rs:597:13
  15:     0x7f04de162fc6 - std::sys_common::backtrace::__rust_end_short_backtrace::h090f3faf8f98a395
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/sys_common/backtrace.rs:151:18
  16:     0x7f04de165a52 - rust_begin_unwind
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/panicking.rs:593:5
  17:     0x7f04de1c59f3 - core::panicking::panic_fmt::h4ec8274704d163a3
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/panicking.rs:67:14
  18:     0x7f04e21268ae - <rustc_transmute[201a2112e531ac6e]::layout::tree::rustc::Err as core[f12ae36cc2e1ecf0]::convert::From<&rustc_middle[4cadb439cfabc8cf]::ty::layout::LayoutError>>::from
  19:     0x7f04e211681f - <rustc_transmute[201a2112e531ac6e]::layout::tree::Tree<rustc_transmute[201a2112e531ac6e]::layout::rustc::Def, rustc_transmute[201a2112e531ac6e]::layout::rustc::Ref>>::from_ty
  20:     0x7f04e211bdc9 - <rustc_transmute[201a2112e531ac6e]::maybe_transmutable::MaybeTransmutableQuery<rustc_middle[4cadb439cfabc8cf]::ty::Ty, rustc_middle[4cadb439cfabc8cf]::ty::context::TyCtxt>>::answer
  21:     0x7f04e2128f85 - <rustc_transmute[201a2112e531ac6e]::rustc::TransmuteTypeEnv>::is_transmutable
  22:     0x7f04dfbbb231 - <rustc_trait_selection[ec0b3e5583b7dbc2]::traits::select::SelectionContext>::confirm_candidate
  23:     0x7f04dfbaee2c - <rustc_trait_selection[ec0b3e5583b7dbc2]::traits::select::SelectionContext>::evaluate_stack
  24:     0x7f04dfbad631 - <rustc_query_system[5a9d202ec1d2890c]::dep_graph::graph::DepGraph<rustc_middle[4cadb439cfabc8cf]::dep_graph::dep_node::DepKind>>::with_anon_task::<rustc_middle[4cadb439cfabc8cf]::ty::context::TyCtxt, <rustc_trait_selection[ec0b3e5583b7dbc2]::traits::select::SelectionContext>::in_task<<rustc_trait_selection[ec0b3e5583b7dbc2]::traits::select::SelectionContext>::evaluate_trait_predicate_recursively::{closure#0}::{closure#2}, core[f12ae36cc2e1ecf0]::result::Result<rustc_middle[4cadb439cfabc8cf]::traits::select::EvaluationResult, rustc_middle[4cadb439cfabc8cf]::traits::select::OverflowError>>::{closure#0}, core[f12ae36cc2e1ecf0]::result::Result<rustc_middle[4cadb439cfabc8cf]::traits::select::EvaluationResult, rustc_middle[4cadb439cfabc8cf]::traits::select::OverflowError>>
  25:     0x7f04e018bb59 - <rustc_trait_selection[ec0b3e5583b7dbc2]::traits::select::SelectionContext>::evaluate_predicate_recursively
  26:     0x7f04e01894c2 - rustc_traits[8460fe50fcddc0b7]::evaluate_obligation::evaluate_obligation
  27:     0x7f04df4344a2 - rustc_query_impl[7f6201a046a7a363]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[7f6201a046a7a363]::query_impl::evaluate_obligation::dynamic_query::{closure#2}::{closure#0}, rustc_middle[4cadb439cfabc8cf]::query::erase::Erased<[u8; 2usize]>>
  28:     0x7f04df93d0f9 - rustc_query_system[5a9d202ec1d2890c]::query::plumbing::try_execute_query::<rustc_query_impl[7f6201a046a7a363]::DynamicConfig<rustc_query_system[5a9d202ec1d2890c]::query::caches::DefaultCache<rustc_middle[4cadb439cfabc8cf]::infer::canonical::Canonical<rustc_middle[4cadb439cfabc8cf]::ty::ParamEnvAnd<rustc_middle[4cadb439cfabc8cf]::ty::Predicate>>, rustc_middle[4cadb439cfabc8cf]::query::erase::Erased<[u8; 2usize]>>, false, false, false>, rustc_query_impl[7f6201a046a7a363]::plumbing::QueryCtxt, false>
  29:     0x7f04df93cd6e - rustc_query_impl[7f6201a046a7a363]::query_impl::evaluate_obligation::get_query_non_incr::__rust_end_short_backtrace
  30:     0x7f04df74f445 - <rustc_trait_selection[ec0b3e5583b7dbc2]::traits::fulfill::FulfillProcessor as rustc_data_structures[cf1e21f3f9509cab]::obligation_forest::ObligationProcessor>::process_obligation
  31:     0x7f04df74a484 - <rustc_data_structures[cf1e21f3f9509cab]::obligation_forest::ObligationForest<rustc_trait_selection[ec0b3e5583b7dbc2]::traits::fulfill::PendingPredicateObligation>>::process_obligations::<rustc_trait_selection[ec0b3e5583b7dbc2]::traits::fulfill::FulfillProcessor>
  32:     0x7f04df7327f9 - <rustc_hir_typeck[f45b206fb7652ba5]::fn_ctxt::FnCtxt>::check_argument_types
  33:     0x7f04df726156 - <rustc_hir_typeck[f45b206fb7652ba5]::fn_ctxt::FnCtxt>::check_call
  34:     0x7f04df67ca56 - <rustc_hir_typeck[f45b206fb7652ba5]::fn_ctxt::FnCtxt>::check_expr_kind
  35:     0x7f04df69197f - <rustc_hir_typeck[f45b206fb7652ba5]::fn_ctxt::FnCtxt>::check_expr_with_expectation
  36:     0x7f04df6c4a61 - <rustc_hir_typeck[f45b206fb7652ba5]::fn_ctxt::FnCtxt>::check_block_with_expected
  37:     0x7f04df67cb9f - <rustc_hir_typeck[f45b206fb7652ba5]::fn_ctxt::FnCtxt>::check_expr_kind
  38:     0x7f04df69197f - <rustc_hir_typeck[f45b206fb7652ba5]::fn_ctxt::FnCtxt>::check_expr_with_expectation
  39:     0x7f04dfd7df2d - <rustc_hir_typeck[f45b206fb7652ba5]::fn_ctxt::FnCtxt>::check_return_expr
  40:     0x7f04dfd75d29 - rustc_hir_typeck[f45b206fb7652ba5]::check::check_fn
  41:     0x7f04dfd5be3b - rustc_hir_typeck[f45b206fb7652ba5]::typeck
  42:     0x7f04df474b7e - rustc_query_impl[7f6201a046a7a363]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[7f6201a046a7a363]::query_impl::typeck::dynamic_query::{closure#2}::{closure#0}, rustc_middle[4cadb439cfabc8cf]::query::erase::Erased<[u8; 8usize]>>
  43:     0x7f04df474b4e - <rustc_query_impl[7f6201a046a7a363]::query_impl::typeck::dynamic_query::{closure#2} as core[f12ae36cc2e1ecf0]::ops::function::FnOnce<(rustc_middle[4cadb439cfabc8cf]::ty::context::TyCtxt, rustc_span[14af2d27fb997609]::def_id::LocalDefId)>>::call_once
  44:     0x7f04df57366d - rustc_query_system[5a9d202ec1d2890c]::query::plumbing::try_execute_query::<rustc_query_impl[7f6201a046a7a363]::DynamicConfig<rustc_query_system[5a9d202ec1d2890c]::query::caches::VecCache<rustc_span[14af2d27fb997609]::def_id::LocalDefId, rustc_middle[4cadb439cfabc8cf]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[7f6201a046a7a363]::plumbing::QueryCtxt, false>
  45:     0x7f04e0de25e1 - rustc_query_impl[7f6201a046a7a363]::query_impl::typeck::get_query_non_incr::__rust_end_short_backtrace
  46:     0x7f04df5f346d - rustc_query_impl[7f6201a046a7a363]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[7f6201a046a7a363]::query_impl::used_trait_imports::dynamic_query::{closure#2}::{closure#0}, rustc_middle[4cadb439cfabc8cf]::query::erase::Erased<[u8; 8usize]>>
  47:     0x7f04df5f33ce - <rustc_query_impl[7f6201a046a7a363]::query_impl::used_trait_imports::dynamic_query::{closure#2} as core[f12ae36cc2e1ecf0]::ops::function::FnOnce<(rustc_middle[4cadb439cfabc8cf]::ty::context::TyCtxt, rustc_span[14af2d27fb997609]::def_id::LocalDefId)>>::call_once
  48:     0x7f04df57366d - rustc_query_system[5a9d202ec1d2890c]::query::plumbing::try_execute_query::<rustc_query_impl[7f6201a046a7a363]::DynamicConfig<rustc_query_system[5a9d202ec1d2890c]::query::caches::VecCache<rustc_span[14af2d27fb997609]::def_id::LocalDefId, rustc_middle[4cadb439cfabc8cf]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[7f6201a046a7a363]::plumbing::QueryCtxt, false>
  49:     0x7f04e0dd8111 - rustc_query_impl[7f6201a046a7a363]::query_impl::used_trait_imports::get_query_non_incr::__rust_end_short_backtrace
  50:     0x7f04e0836835 - rustc_hir_analysis[27a0f02ea330845c]::check_unused::check_crate
  51:     0x7f04e0833ca1 - rustc_hir_analysis[27a0f02ea330845c]::check_crate
  52:     0x7f04e082d22a - rustc_interface[7e6a9899d53a1fe5]::passes::analysis
  53:     0x7f04e087dcfa - rustc_query_impl[7f6201a046a7a363]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[7f6201a046a7a363]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[4cadb439cfabc8cf]::query::erase::Erased<[u8; 1usize]>>
  54:     0x7f04e087dce9 - <rustc_query_impl[7f6201a046a7a363]::query_impl::analysis::dynamic_query::{closure#2} as core[f12ae36cc2e1ecf0]::ops::function::FnOnce<(rustc_middle[4cadb439cfabc8cf]::ty::context::TyCtxt, ())>>::call_once
  55:     0x7f04e0ab1318 - rustc_query_system[5a9d202ec1d2890c]::query::plumbing::try_execute_query::<rustc_query_impl[7f6201a046a7a363]::DynamicConfig<rustc_query_system[5a9d202ec1d2890c]::query::caches::SingleCache<rustc_middle[4cadb439cfabc8cf]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[7f6201a046a7a363]::plumbing::QueryCtxt, false>
  56:     0x7f04e0ab10e9 - rustc_query_impl[7f6201a046a7a363]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  57:     0x7f04e092ddf2 - <rustc_interface[7e6a9899d53a1fe5]::queries::QueryResult<&rustc_middle[4cadb439cfabc8cf]::ty::context::GlobalCtxt>>::enter::<core[f12ae36cc2e1ecf0]::result::Result<(), rustc_span[14af2d27fb997609]::ErrorGuaranteed>, rustc_driver_impl[10725d833993dc31]::run_compiler::{closure#1}::{closure#2}::{closure#4}>
  58:     0x7f04e092ca97 - <rustc_interface[7e6a9899d53a1fe5]::interface::Compiler>::enter::<rustc_driver_impl[10725d833993dc31]::run_compiler::{closure#1}::{closure#2}, core[f12ae36cc2e1ecf0]::result::Result<core[f12ae36cc2e1ecf0]::option::Option<rustc_interface[7e6a9899d53a1fe5]::queries::Linker>, rustc_span[14af2d27fb997609]::ErrorGuaranteed>>
  59:     0x7f04e0929cc5 - rustc_span[14af2d27fb997609]::set_source_map::<core[f12ae36cc2e1ecf0]::result::Result<(), rustc_span[14af2d27fb997609]::ErrorGuaranteed>, rustc_interface[7e6a9899d53a1fe5]::interface::run_compiler<core[f12ae36cc2e1ecf0]::result::Result<(), rustc_span[14af2d27fb997609]::ErrorGuaranteed>, rustc_driver_impl[10725d833993dc31]::run_compiler::{closure#1}>::{closure#0}::{closure#0}>
  60:     0x7f04e0929736 - <scoped_tls[a7f541bbfecfca9d]::ScopedKey<rustc_span[14af2d27fb997609]::SessionGlobals>>::set::<rustc_interface[7e6a9899d53a1fe5]::interface::run_compiler<core[f12ae36cc2e1ecf0]::result::Result<(), rustc_span[14af2d27fb997609]::ErrorGuaranteed>, rustc_driver_impl[10725d833993dc31]::run_compiler::{closure#1}>::{closure#0}, core[f12ae36cc2e1ecf0]::result::Result<(), rustc_span[14af2d27fb997609]::ErrorGuaranteed>>
  61:     0x7f04e0928cfc - std[7c7acd4e376d60d3]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[7e6a9899d53a1fe5]::util::run_in_thread_pool_with_globals<rustc_interface[7e6a9899d53a1fe5]::interface::run_compiler<core[f12ae36cc2e1ecf0]::result::Result<(), rustc_span[14af2d27fb997609]::ErrorGuaranteed>, rustc_driver_impl[10725d833993dc31]::run_compiler::{closure#1}>::{closure#0}, core[f12ae36cc2e1ecf0]::result::Result<(), rustc_span[14af2d27fb997609]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[f12ae36cc2e1ecf0]::result::Result<(), rustc_span[14af2d27fb997609]::ErrorGuaranteed>>
  62:     0x7f04e0ca1305 - <<std[7c7acd4e376d60d3]::thread::Builder>::spawn_unchecked_<rustc_interface[7e6a9899d53a1fe5]::util::run_in_thread_pool_with_globals<rustc_interface[7e6a9899d53a1fe5]::interface::run_compiler<core[f12ae36cc2e1ecf0]::result::Result<(), rustc_span[14af2d27fb997609]::ErrorGuaranteed>, rustc_driver_impl[10725d833993dc31]::run_compiler::{closure#1}>::{closure#0}, core[f12ae36cc2e1ecf0]::result::Result<(), rustc_span[14af2d27fb997609]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[f12ae36cc2e1ecf0]::result::Result<(), rustc_span[14af2d27fb997609]::ErrorGuaranteed>>::{closure#1} as core[f12ae36cc2e1ecf0]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  63:     0x7f04de170435 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hc0b1022758ecac73
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/alloc/src/boxed.rs:1993:9
  64:     0x7f04de170435 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h0c9654ebe7ad657e
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/alloc/src/boxed.rs:1993:9
  65:     0x7f04de170435 - std::sys::unix::thread::Thread::new::thread_start::h04c8e9c7d83d3bd5
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/sys/unix/thread.rs:108:17
  66:     0x7f04dde8c9eb - <unknown>
  67:     0x7f04ddf10ebc - <unknown>
  68:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.72.0 (5680fa18f 2023-08-23) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [evaluate_obligation] evaluating trait selection obligation `(): core::mem::transmutability::BikeshedIntrinsicFrom<should_pad_explicitly_packed_field::ExplicitlyPadded, {type error}, core::mem::transmutability::Assume { alignment: false, lifetimes: false, safety: false, validity: false }>`
#1 [typeck] type-checking `should_pad_explicitly_packed_field`
#2 [used_trait_imports] finding used_trait_imports `should_pad_explicitly_packed_field`
#3 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 5 previous errors; 1 warning emitted

Some errors have detailed explanations: E0412, E0601, E0658.
For more information about an error, try `rustc --explain E0412`.
```

</p>
</details>

--- a/compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs
+++ b/compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs
@@ -2920,6 +2920,16 @@ impl<'tcx> InferCtxtPrivExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                     rustc_transmute::Reason::DstIsTooBig => {
                         format!("The size of `{src}` is smaller than the size of `{dst}`")
                     }
+                    rustc_transmute::Reason::SrcSizeOverflow => {
+                        format!(
+                            "values of the type `{src}` are too big for the current architecture"
+                        )
+                    }
+                    rustc_transmute::Reason::DstSizeOverflow => {
+                        format!(
+                            "values of the type `{dst}` are too big for the current architecture"
+                        )
+                    }
                     rustc_transmute::Reason::DstHasStricterAlignment {
                         src_min_align,
                         dst_min_align,

--- a/compiler/rustc_transmute/src/layout/tree.rs
+++ b/compiler/rustc_transmute/src/layout/tree.rs
@@ -189,6 +189,8 @@ pub(crate) mod rustc {
         Unspecified,
         /// This error will be surfaced elsewhere by rustc, so don't surface it.
         UnknownLayout,
+        /// Overflow size
+        SizeOverflow,
         TypeError(ErrorGuaranteed),
     }
 
@@ -196,6 +198,7 @@ pub(crate) mod rustc {
         fn from(err: &LayoutError<'tcx>) -> Self {
             match err {
                 LayoutError::Unknown(..) | LayoutError::ReferencesError(..) => Self::UnknownLayout,
+                LayoutError::SizeOverflow(..) => Self::SizeOverflow,
                 err => unimplemented!("{:?}", err),
             }
         }

--- a/compiler/rustc_transmute/src/lib.rs
+++ b/compiler/rustc_transmute/src/lib.rs
@@ -64,6 +64,10 @@ pub enum Reason {
     SrcLayoutUnknown,
     /// The layout of dst is unknown
     DstLayoutUnknown,
+    /// The size of src is overflow
+    SrcSizeOverflow,
+    /// The size of dst is overflow
+    DstSizeOverflow,
 }
 
 #[cfg(feature = "rustc")]

--- a/compiler/rustc_transmute/src/maybe_transmutable/mod.rs
+++ b/compiler/rustc_transmute/src/maybe_transmutable/mod.rs
@@ -85,6 +85,8 @@ mod rustc {
                 (_, Err(Err::UnknownLayout)) => Answer::No(Reason::DstLayoutUnknown),
                 (Err(Err::Unspecified), _) => Answer::No(Reason::SrcIsUnspecified),
                 (_, Err(Err::Unspecified)) => Answer::No(Reason::DstIsUnspecified),
+                (Err(Err::SizeOverflow), _) => Answer::No(Reason::SrcSizeOverflow),
+                (_, Err(Err::SizeOverflow)) => Answer::No(Reason::DstSizeOverflow),
                 (Ok(src), Ok(dst)) => {
                     MaybeTransmutableQuery { src, dst, scope, assume, context }.answer()
                 }

--- a/tests/ui/transmute/issue-115402-overflow-size.rs
+++ b/tests/ui/transmute/issue-115402-overflow-size.rs
@@ -1,0 +1,27 @@
+#![crate_type = "lib"]
+#![feature(transmutability)]
+mod assert {
+    use std::mem::BikeshedIntrinsicFrom;
+    struct Context;
+
+    pub fn is_maybe_transmutable<Src, Dst>()
+    where
+        Dst: BikeshedIntrinsicFrom<Src, Context>,
+    {
+    }
+}
+
+fn main() {
+    pub union Uninit {
+        a: [u8; usize::MAX],
+    }
+
+    #[repr(C)]
+    struct ExplicitlyPadded(Uninit);
+
+    assert::is_maybe_transmutable::<(), ExplicitlyPadded>();
+    //~^ ERROR `()` cannot be safely transmuted into `ExplicitlyPadded`
+
+    assert::is_maybe_transmutable::<ExplicitlyPadded, ()>();
+    //~^ ERROR `ExplicitlyPadded` cannot be safely transmuted into `()`
+}

--- a/tests/ui/transmute/issue-115402-overflow-size.stderr
+++ b/tests/ui/transmute/issue-115402-overflow-size.stderr
@@ -1,0 +1,33 @@
+error[E0277]: `()` cannot be safely transmuted into `ExplicitlyPadded` in the defining scope of `assert::Context`
+  --> $DIR/issue-115402-overflow-size.rs:22:41
+   |
+LL |     assert::is_maybe_transmutable::<(), ExplicitlyPadded>();
+   |                                         ^^^^^^^^^^^^^^^^ values of the type `ExplicitlyPadded` are too big for the current architecture
+   |
+note: required by a bound in `is_maybe_transmutable`
+  --> $DIR/issue-115402-overflow-size.rs:9:14
+   |
+LL |     pub fn is_maybe_transmutable<Src, Dst>()
+   |            --------------------- required by a bound in this function
+LL |     where
+LL |         Dst: BikeshedIntrinsicFrom<Src, Context>,
+   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `is_maybe_transmutable`
+
+error[E0277]: `ExplicitlyPadded` cannot be safely transmuted into `()` in the defining scope of `assert::Context`
+  --> $DIR/issue-115402-overflow-size.rs:25:55
+   |
+LL |     assert::is_maybe_transmutable::<ExplicitlyPadded, ()>();
+   |                                                       ^^ values of the type `ExplicitlyPadded` are too big for the current architecture
+   |
+note: required by a bound in `is_maybe_transmutable`
+  --> $DIR/issue-115402-overflow-size.rs:9:14
+   |
+LL |     pub fn is_maybe_transmutable<Src, Dst>()
+   |            --------------------- required by a bound in this function
+LL |     where
+LL |         Dst: BikeshedIntrinsicFrom<Src, Context>,
+   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `is_maybe_transmutable`
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0277`.

